engines/wintermute/base/gfx/opengl/shadow_volume.cpp

```diff
@@ -52,141 +55,137 @@ void ShadowVolume::addVertex(Math::Vector3d &vertex) {
-	//	Rend->m_Device->SetTextureStageState(0, D3DTSS_ALPHAARG1, D3DTA_TEXTURE);
-	//	Rend->m_Device->SetTextureStageState(0, D3DTSS_ALPHAARG2, D3DTA_DIFFUSE);
-	//	Rend->m_Device->SetTextureStageState(0, D3DTSS_ALPHAOP,   D3DTOP_MODULATE);
+	glDisable(GL_FOG);
+	glDisable(GL_LIGHTING);
+	glDisable(GL_ALPHA_TEST);
 
-	//	// Only write where stencil val >= 1 (count indicates # of shadows that
-	//	// overlap that pixel)
-	//	Rend->m_Device->SetRenderState(D3DRS_STENCILREF,  0x1 );
-	//	Rend->m_Device->SetRenderState(D3DRS_STENCILFUNC, D3DCMP_LESSEQUAL);
-	//	Rend->m_Device->SetRenderState(D3DRS_STENCILPASS, D3DSTENCILOP_KEEP);
+	_gameRef->_renderer3D->setProjection2D();
 
-	//	Rend->m_Device->SetRenderState(D3DRS_FOGENABLE,          FALSE);
+	glBindTexture(GL_TEXTURE_2D, 0);
+	glBindBuffer(GL_ARRAY_BUFFER, 0);
 
-	//	// Draw a big, gray square
-	//	Rend->m_Device->SetRenderState(D3DRS_COLORVERTEX, TRUE);
+	glEnableClientState(GL_COLOR_ARRAY);
+	glEnableClientState(GL_VERTEX_ARRAY);
+	glDisableClientState(GL_TEXTURE_COORD_ARRAY);
 
-	//	C3DUtils::SetFixedVertexShader(Rend->m_Device, D3DFVF_SHADOWVOLVERTEX);
-	//	C3DUtils::SetStreamSource(Rend->m_Device, 0, m_StencilMaskVB, sizeof(SHADOWVOLVERTEX));
-	//	Rend->m_Device->DrawPrimitive(D3DPT_TRIANGLESTRIP, 0, 2);
+	// Draw a big, gray square
+	glInterleavedArrays(GL_C4UB_V3F, 0, _shadowMask);
+	glDrawArrays(GL_TRIANGLE_STRIP, 0, 4);
 
-	//	// Restore render states
-	//	Rend->m_Device->SetRenderState(D3DRS_ZENABLE,          TRUE);
-	//	Rend->m_Device->SetRenderState(D3DRS_STENCILENABLE,    FALSE);
+	// Restore render states
+	glEnable(GL_DEPTH_TEST);
+	glDisable(GL_STENCIL_TEST);
 
-	//	Rend->Setup3D(NULL, true);
+	_gameRef->_renderer3D->setup3D(nullptr, true);
 
-	//	// clear stencil buffer
-	//	Rend->m_Device->Clear(0, NULL, D3DCLEAR_STENCIL, 0x00, 1.0f, 0);
+	// clear stencil buffer
+	glClearStencil(0);
+	glClear(GL_STENCIL_BUFFER_BIT);
 
 	return true;
 }
 
 //////////////////////////////////////////////////////////////////////////
 bool ShadowVolume::initMask() {
-	warning("ShadowVolume::initMask not implemented yet");
-	//	CBRenderD3D* Rend = (CBRenderD3D*)Game->m_Renderer;
-
-	//	RELEASE(m_StencilMaskVB);
-
-	//	Rend->m_Device->CreateVertexBufferUni(4 * sizeof(SHADOWVOLVERTEX),	D3DUSAGE_WRITEONLY, D3DFVF_SHADOWVOLVERTEX,	D3DPOOL_MANAGED, &m_StencilMaskVB);
-	//	SHADOWVOLVERTEX* v;
-	//	m_StencilMaskVB->LockVB(0, 0, &v, 0);
-	//	{
-	//		v[0].p = D3DXVECTOR4(0.0f, Rend->m_RealHeight, 0.0f, 1.0f);
-	//		v[1].p = D3DXVECTOR4(0.0f, 0.0f, 0.0f, 1.0f);
-	//		v[2].p = D3DXVECTOR4(Rend->m_RealWidth, Rend->m_RealHeight, 0.0f, 1.0f);
-	//		v[3].p = D3DXVECTOR4(Rend->m_RealWidth, 0.0f, 0.0f, 1.0f);
-	//		v[0].color = m_Color;
-	//		v[1].color = m_Color;
-	//		v[2].color = m_Color;
-	//		v[3].color = m_Color;
-	//	}
-	//	m_StencilMaskVB->Unlock();
+	Rect32 viewport = _gameRef->_renderer->getViewPort();
+
+	_shadowMask[0].x = viewport.left;
```

Comment:
Can perhaps be simplified with constructor/initializer-syntax, but given the amount of fields, and that C++ doesn't have the same nice struct-initializer syntax as C, perhaps not.

Reply:
I would prefer it as is over the initializer syntax, at least in this case